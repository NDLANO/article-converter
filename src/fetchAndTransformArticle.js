import cheerio from 'cheerio';
import { fetchArticle } from './api/articleApi';
import { transform } from './transformers';
import config from './config';
import { getCopyString } from './plugins/pluginHelpers';

export async function transformArticle(
  article,
  lang,
  accessToken,
  options = {}
) {
  const articleContent = article.content.content
    ? cheerio.load(article.content.content, {
        recognizeSelfClosing: true,
      })
    : undefined;
  const { html, embedMetaData } = articleContent
    ? await transform(
        articleContent,
        lang,
        accessToken,
        article.visualElement,
        options
      )
    : {};
  const copyText = getCopyString(
    article.title.title,
    config.ndlaFrontendDomain,
    options.path,
    article.copyright,
    lang
  );

  return {
    ...article,
    content: html || '',
    metaData: { ...embedMetaData, copyText } || '',
    title: article.title.title || '',
    tags: article.tags ? article.tags.tags : [],
    introduction: article.introduction
      ? article.introduction.introduction
      : undefined,
    metaDescription: article.metaDescription
      ? article.metaDescription.metaDescription
      : '',
  };
}

export default async function fetchAndTransformArticle(
  articleId,
  lang,
  accessToken,
  options = {}
) {
  const article = await fetchArticle(articleId, accessToken, lang);
  const transformedArticle = await transformArticle(
    article,
    lang,
    accessToken,
    options
  );
  return transformedArticle;
}
