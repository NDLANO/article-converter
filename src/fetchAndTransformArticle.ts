import cheerio from 'cheerio';
import { IArticleV2 } from '@ndla/types-article-api';
import { fetchArticle } from './api/articleApi';
import { transform } from './transformers';
import config from './config';
import { getCopyString } from './plugins/pluginHelpers';
import { LocaleType, ResponseHeaders, TransformedArticle, TransformOptions } from './interfaces';

export async function transformArticle(
  article: IArticleV2,
  headers: ResponseHeaders,
  lang: LocaleType,
  accessToken: string,
  feideToken: string,
  options: TransformOptions = {},
): Promise<TransformedArticle> {
  const articleContent = article.content.content
    ? cheerio.load(article.content.content, {
        recognizeSelfClosing: true,
      })
    : undefined;

  const { html, embedMetaData, responseHeaders } = articleContent
    ? await transform(
        articleContent,
        headers,
        lang,
        accessToken,
        feideToken,
        article.visualElement,
        options,
      )
    : { html: undefined, embedMetaData: undefined, responseHeaders: undefined };

  const htmlString: string = html ?? '';

  const copyText: string = getCopyString(
    article.title.title,
    config.ndlaFrontendDomain,
    options.path,
    article.copyright,
    lang,
  );
  const hasContent = article.articleType === 'standard' || cheerio.load(htmlString).text() !== '';

  return {
    ...article,
    content: hasContent ? htmlString : '',
    metaData: { ...embedMetaData, copyText } || '',
    title: article.title.title || '',
    tags: article.tags ? article.tags.tags : [],
    introduction: article.introduction ? article.introduction.introduction : undefined,
    metaDescription: article.metaDescription ? article.metaDescription.metaDescription : '',
    headerData: responseHeaders ?? {},
  };
}

export default async function fetchAndTransformArticle(
  articleId: string,
  lang: LocaleType,
  accessToken: string,
  feideToken: string,
  options = {},
): Promise<TransformedArticle> {
  const [article, headers] = await fetchArticle(articleId, accessToken, feideToken, lang);
  return await transformArticle(article, headers, lang, accessToken, feideToken, options);
}
