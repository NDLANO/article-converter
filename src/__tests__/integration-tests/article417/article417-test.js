/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import bunyan from 'bunyan';
import { prettify } from '../../testHelpers';
import article417 from './article417';
import articleResource413 from './articleResource413';
import articleResource414 from './articleResource414';
import fetchAndTransformArticle from '../../../fetchAndTransformArticle';
import log from '../../../utils/logger';

const resources = {
  '413': articleResource413,
  '414': articleResource414,
  '416': articleResource413,
};

test('app/fetchAndTransformArticle 417', async () => {
  nock('http://ndla-api')
    .get('/article-api/v2/articles/417?language=nb&fallback=true')
    .reply(200, article417);
  ['413', '414', '416'].forEach(id => {
    nock('http://ndla-api')
      .get(`/article-api/v2/articles/${id}?language=nb&fallback=true`)
      .reply(200, article417);
    nock('http://ndla-api')
      .get(
        `/taxonomy/v1/queries/resources?contentURI=urn:article:${id}&language=nb`
      )
      .reply(200, resources[id]);
  });

  log.level(bunyan.FATAL + 1); // temporarily disable logging

  const transformed = await fetchAndTransformArticle('417', 'nb', 'some_token');
  const { content, ...rest } = transformed;

  log.level(bunyan.INFO);

  expect(rest).toMatchSnapshot();
  expect(prettify(content)).toMatchSnapshot();
});
