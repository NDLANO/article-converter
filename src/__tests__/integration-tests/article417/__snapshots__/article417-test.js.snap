// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/fetchAndTransformArticle 417 1`] = `
Object {
  "articleType": "standard",
  "copyright": Object {
    "creators": Array [
      Object {
        "name": "Johannes Leiknes Nag",
        "type": "Writer",
      },
    ],
    "license": Object {
      "description": "Creative Commons Attribution-ShareAlike 2.0 Generic",
      "license": "by-sa",
      "url": "https://creativecommons.org/licenses/by-sa/2.0/",
    },
    "origin": "",
    "processors": Array [],
    "rightsholders": Array [],
  },
  "created": "2016-03-22T08:46:42Z",
  "id": 417,
  "introduction": "Vi bruker skript for å tilføre funksjonalitet og interaktivitet på nettsider utover det vi kan lage med markeringsspråk og stilark. Det kan være alt fra små justeringer i design til hele spillkonsept.",
  "metaData": Object {},
  "metaDescription": "Vi bruker skript for å tilføre funksjonalitet og interaktivitet på nettsider utover det vi kan lage med markeringsspråk og stilark. Det kan være alt fra små justeringer i design til hele spillkonsept.",
  "oldNdlaUrl": "//red.ndla.no/node/160589",
  "requiredLibraries": Array [],
  "revision": 1,
  "supportedLanguages": Array [
    "nb",
    "nn",
  ],
  "tags": Array [
    "design",
    "koding",
    "nettside",
    "programmering",
    "tagger",
  ],
  "title": "Skripting gjør sidene mer interaktive",
  "updated": "2017-03-07T17:16:47Z",
  "updatedBy": "content-import-client",
}
`;

exports[`app/fetchAndTransformArticle 417 2`] = `
"

<section>

  <figure class=\\"c-figure\\">

    <img alt=\\"En feil oppsto ved lasting av en ekstern ressurs.\\"
      src=\\"data:image/svg+xml;charset=UTF-8,
      <svg fill=\\"#8A8888\\" height=\\"400\\" viewBox=\\"0 0 24 12\\" width=\\"100%\\" xmlns=\\"http://www.w3.org/2000/svg\\" style=\\"background-color: #EFF0F2\\">
        <path d=\\"M0 0h24v24H0V0z\\" fill=\\"none\\"/>
        <path transform=\\"scale(0.3) translate(28, 8.5)\\" d=\\"M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z\\"/>
      </svg>
    \\">

    <figcaption>En feil oppsto ved lasting av en ekstern ressurs.</figcaption>
  </figure>
</section>
<section>
  <p>Vil du lære deg å programmere? Da må du først velge et programmeringsspråk. Jobber du med nettsider, er det naturlig å lære seg JavaScript. Det er flere måter å komme i gang på. Du kan for eksempel følge et nettbasert kurs som det vi ser i videoen over, eller opplegget hos
    <a href=\\"https://nb.khanacademy.org/computing/computer-programming\\"
      rel=\\"noopener noreferrer\\"
      target=\\"_blank\\"
      title=\\"Khan Academy\\">Khan Academy
    </a>. En annen tilnærming er å finne eksempelkoder som man tar i bruk på egne websider – vi kan kalle det problembasert tilnærming. Mange vil oppleve det svært nyttig.
  </p>
  <blockquote>«Everyone should learn how to program a computer because it teaches you how to think.» — Steve Jobs
  </blockquote>
  <p>Det finnes mange kodespråk. Noen av dem bygger kode med byggeklosser i stedet for ord – slik som
    <a href=\\"https://scratch.mit.edu/projects/editor/?tip_bar=getStarted\\"
      rel=\\"noopener noreferrer\\"
      target=\\"_blank\\"
      title=\\"Scratch på norsk\\">Scratch på norsk
    </a>.
  </p>
  <h2>Endre innhold</h2>
  <p>Vi kan ha behov for å endre innhold på nettsidene våre basert på hva brukeren gjør. For enkelhets skyld viser vi her skript direkte i HTML-koden:
  </p>
  <pre>
    <code><h1
      <strong>onclick
      </strong>=\\"innerHTML='Lengre og mer utførlig overskrift'\\">Kort overskrift</h1>
      <br>

    </code>
  </pre>
  <p>Hvis man klikker på overskriften «Kort overskrift», blir den byttet ut med en annen overskrift fordi skriptet endrer verdien innerHTML på H1-taggen.
  </p>
  <p>Vanligvis samles skriptene i et eget dokument, på samme måte som stilark. Vi kan da lage funksjonalitet som kan gjenbrukes i alle dokumenter, f.eks. innholdsfortegnelser:
  </p>
  <pre>
    <code>var overskrifter = document.getElementByTagName(\\"h1\\");
      <br>

    </code>
  </pre>
  <p>Her vil alle overskrifter fanges opp og lagres, slik at de senere kan vises fram et annet sted – eller på forespørsel.
  </p>
  <p>I eksempelet over er
    <em>overskrifter</em>en
    <strong>variabel</strong>og
    <em>getElementByTagName</em>en
    <strong>funksjon
    </strong>. En funksjon beregner eller gjør noe, mens en variabel lagrer eller husker for oss.
  </p>
  <h2>Endre stiler</h2>
  <p>Å endre layout, synlighet og design er enda mer vanlig:</p>
  <pre>
    <code><h1
      <strong>onclick</strong>=\\"style='background:yellow'\\">Min overskrift</h1>
      <br>

    </code>
  </pre>
  <p>Med denne koden får overskriften gul bakgrunnsfarge. Med stiler kan vi endre både synlighet, størrelse, animasjon og annet.
  </p>
  <h2>Validering</h2>
  <p>Mange nettsider har ulike skjema. Det kan være for søk eller innsending av forespørsler. Her kan det fort bli feiltasting siden brukeren er fri til å skrive inn det han ønsker. Skript vil her komme til nytte både for å veilede brukeren og for å sikre at riktig informasjon blir utfylt.
  </p>
  <p>For å sjekke et telefonnummer trengs et skript som sjekker:</p>
  <ul>
    <li>om det kun er skrevet inn tall</li>
    <li>om det er åtte siffer</li>
  </ul>
  <p>Skriptet kan da gi en feilmelding om «Feilskrevet telefonnummer» eller en mer detaljert feilmelding som «Du har skrevet tegn som ikke er tall». Sannsynligvis burde skriptet være så komplekst at det tok høyde for valgfri landskode (f.eks. +47).
  </p>
  <p>Vi kan bruke valg-setninger for å teste:</p>
  <pre>
    <code>
      <strong>if</strong>(telefonFelt.value.length != 8) { /* != betyr ulik */
      <br>
      alert(\\"Du må skrive inn åtte tall.\\");
      <br>
      }
    </code>
  </pre>
  <p>Bruk av
    <strong>valg
    </strong>er en av de store styrkene til et programmeringsspråk, noe du fort vil erfare. Programmering av
    <strong>gjentakelser
    </strong>med løkker er en annen. Vi kan sette opp en løkke som for eksempel repeterer samme testkode for alle felt i skjemaet vårt.
  </p>
</section>
<section>
  <section class=\\"c-related-articles\\">
    <h1 class=\\"c-related-articles__component-title\\">Relaterte artikler</h1>
    <div class=\\"c-related-articles__articles\\">
      <article class=\\"c-related-articles__item c-related-articles__item--tasks-and-activities\\">
        <h1 class=\\"c-related-articles__title\\">
          <div class=\\"c-content-type-badge c-content-type-badge--tasks-and-activities c-content-type-badge--small c-content-type-badge--background\\">
            <svg fill=\\"currentColor\\"
              preserveAspectRatio=\\"xMidYMid meet\\"
              height=\\"1em\\"
              width=\\"1em\\"
              class=\\"c-icon\\"
              viewBox=\\"0 0 24 24\\"
              data-license=\\"Apache License 2.0\\"
              data-source=\\"Material Design\\"
              style=\\"vertical-align:middle\\">
              <g>
                <title>Oppgaver og aktiviteter</title>
                <g id=\\"Page-1\\"
                  stroke=\\"none\\"
                  fill=\\"none\\"
                  stroke-width=\\"1\\"
                  fill-rule=\\"evenodd\\">
                  <g id=\\"oppgaver-og-aktiviteter\\">
                    <g id=\\"Shape\\">
                      <polygon points=\\"0 0 24 0 24 24 0 24\\" />

                      <path d=\\"M17.75,8 L14,4.25 L4,14.25 L4,18 L7.75,18 L17.75,8 Z M20.71,5.04 C21.1,4.65 21.1,4.02 20.71,3.63 L18.37,1.29 C17.98,0.9 17.35,0.9 16.96,1.29 L15,3.25 L18.75,7 L20.71,5.04 Z\\"
                        fill=\\"currentColor\\"
                        fill-rule=\\"nonzero\\" />

                      <polygon fill=\\"currentColor\\"
                        points=\\"0 22 24 22 24 24 0 24\\"
                        fill-opacity=\\"0.36\\"
                        fill-rule=\\"nonzero\\" />

                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </div>
          <a href=\\"/subjects/subject:12/topic:1:183846/topic:1:183935/topic:1:110265/resource:1:110271\\"
            class=\\"c-related-articles__link\\">Skripting gjør sidene mer interaktive
          </a>
        </h1>
        <p class=\\"c-related-articles__description\\">Vi bruker skript for å tilføre funksjonalitet og interaktivitet på nettsider utover det vi kan lage med markeringsspråk og stilark. Det kan være alt fra små justeringer i design til hele spillkonsept.
        </p>
      </article>
      <article class=\\"c-related-articles__item c-related-articles__item--subject-material\\">
        <h1 class=\\"c-related-articles__title\\">
          <div class=\\"c-content-type-badge c-content-type-badge--subject-material c-content-type-badge--small c-content-type-badge--background\\">
            <svg fill=\\"currentColor\\"
              preserveAspectRatio=\\"xMidYMid meet\\"
              height=\\"1em\\"
              width=\\"1em\\"
              class=\\"c-icon\\"
              viewBox=\\"0 0 24 24\\"
              data-license=\\"Apache License 2.0\\"
              data-source=\\"Material Design\\"
              style=\\"vertical-align:middle\\">
              <g>
                <title>Fagstoff</title>
                <g id=\\"Page-1\\"
                  stroke=\\"none\\"
                  fill=\\"none\\"
                  stroke-width=\\"1\\"
                  fill-rule=\\"evenodd\\">
                  <g id=\\"fagstoff\\">
                    <g>
                      <polygon id=\\"Shape\\"
                        points=\\"0 0 24 0 24 24 0 24\\" />

                      <path d=\\"M0,10 L0,8 L11,8 L11,10 L0,10 Z M24,0 L24,2 L0,2 L0,0 L24,0 Z M24,4 L24,6 L0,6 L0,4 L24,4 Z M13,8 L24,15.5 L13,23 L13,8 Z\\"
                        id=\\"shape\\"
                        fill=\\"currentColor\\" />

                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </div>
          <a href=\\"/subjects/subject:1/topic:1:172733/topic:1:169379/resource:1:158547\\"
            class=\\"c-related-articles__link\\">Skripting gjør sidene mer interaktive
          </a>
        </h1>
        <p class=\\"c-related-articles__description\\">Vi bruker skript for å tilføre funksjonalitet og interaktivitet på nettsider utover det vi kan lage med markeringsspråk og stilark. Det kan være alt fra små justeringer i design til hele spillkonsept.
        </p>
      </article>
      <article class=\\"c-related-articles__item c-related-articles__item--tasks-and-activities c-related-articles__item--hidden\\">
        <h1 class=\\"c-related-articles__title\\">
          <div class=\\"c-content-type-badge c-content-type-badge--tasks-and-activities c-content-type-badge--small c-content-type-badge--background\\">
            <svg fill=\\"currentColor\\"
              preserveAspectRatio=\\"xMidYMid meet\\"
              height=\\"1em\\"
              width=\\"1em\\"
              class=\\"c-icon\\"
              viewBox=\\"0 0 24 24\\"
              data-license=\\"Apache License 2.0\\"
              data-source=\\"Material Design\\"
              style=\\"vertical-align:middle\\">
              <g>
                <title>Oppgaver og aktiviteter</title>
                <g id=\\"Page-1\\"
                  stroke=\\"none\\"
                  fill=\\"none\\"
                  stroke-width=\\"1\\"
                  fill-rule=\\"evenodd\\">
                  <g id=\\"oppgaver-og-aktiviteter\\">
                    <g id=\\"Shape\\">
                      <polygon points=\\"0 0 24 0 24 24 0 24\\" />

                      <path d=\\"M17.75,8 L14,4.25 L4,14.25 L4,18 L7.75,18 L17.75,8 Z M20.71,5.04 C21.1,4.65 21.1,4.02 20.71,3.63 L18.37,1.29 C17.98,0.9 17.35,0.9 16.96,1.29 L15,3.25 L18.75,7 L20.71,5.04 Z\\"
                        fill=\\"currentColor\\"
                        fill-rule=\\"nonzero\\" />

                      <polygon fill=\\"currentColor\\"
                        points=\\"0 22 24 22 24 24 0 24\\"
                        fill-opacity=\\"0.36\\"
                        fill-rule=\\"nonzero\\" />

                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </div>
          <a href=\\"/subjects/subject:12/topic:1:183846/topic:1:183935/topic:1:110265/resource:1:110271\\"
            class=\\"c-related-articles__link\\">Skripting gjør sidene mer interaktive
          </a>
        </h1>
        <p class=\\"c-related-articles__description\\">Vi bruker skript for å tilføre funksjonalitet og interaktivitet på nettsider utover det vi kan lage med markeringsspråk og stilark. Det kan være alt fra små justeringer i design til hele spillkonsept.
        </p>
      </article>
    </div>
    <button class=\\"c-button c-button--outline c-related-articles__button\\"
      data-showmore=\\"Vis flere relaterte artikler\\"
      data-showless=\\"Vis mindre\\"
      type=\\"button\\">Vis flere relaterte artikler
    </button>
  </section>
</section>"
`;
